File: src/test/java/io/lsdconsulting/lsd/distributed/generator/diagram/LsdLoggerShould.java
package io.lsdconsulting.lsd.distributed.generator.diagram;

import com.lsd.LsdContext;
import com.lsd.events.Message;
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction;
import io.lsdconsulting.lsd.distributed.access.repository.InterceptedDocumentRepository;
import io.lsdconsulting.lsd.distributed.generator.diagram.dto.EventContainer;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;
import java.util.Optional;

import static java.util.Collections.singletonList;
import static org.apache.commons.lang3.RandomStringUtils.randomAlphanumeric;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.*;

public class LsdLoggerShould {

    private final InterceptedDocumentRepository interceptedDocumentRepository = mock(InterceptedDocumentRepository.class);
    private final InteractionGenerator interactionGenerator = mock(InteractionGenerator.class);
    private final LsdContext lsdContext = mock(LsdContext.class);

    private final String traceId = randomAlphanumeric(8);
    private final String secondaryTraceId = randomAlphanumeric(8);

    private final LsdLogger underTest = new LsdLogger(interactionGenerator, lsdContext);

    @Test
    public void captureInteractionName() {
        final InterceptedInteraction interceptedInteraction = InterceptedInteraction.builder().build();
        given(interceptedDocumentRepository.findByTraceIds(traceId)).willReturn(singletonList(interceptedInteraction));
        Message message = Message.builder().label("interactionName").data("body").build();
        given(interactionGenerator.generate(any())).willReturn(EventContainer.builder().events(singletonList(message)).build());

        underTest.captureInteractionsFromDatabase(traceId);

        verify(interactionGenerator).generate(Map.of(traceId, Optional.empty()));
        verify(lsdContext).capture(message);
    }

    @Test
    public void captureInteractionNamesWithColour() {
        final InterceptedInteraction interceptedInteraction1 = InterceptedInteraction.builder().build();
        final InterceptedInteraction interceptedInteraction2 = InterceptedInteraction.builder().build();
        given(interceptedDocumentRepository.findByTraceIds(traceId)).willReturn(singletonList(interceptedInteraction1));
        given(interceptedDocumentRepository.findByTraceIds(secondaryTraceId)).willReturn(singletonList(interceptedInteraction2));
        Message message = Message.builder().label("interactionName").data("body").build();
        given(interactionGenerator.generate(any())).willReturn(EventContainer.builder().events(List.of(message, message)).build());

        underTest.captureInteractionsFromDatabase(Map.of(traceId, Optional.of("red"), secondaryTraceId, Optional.of("green")));

        verify(interactionGenerator).generate(Map.of(traceId, Optional.of("red"), secondaryTraceId, Optional.of("green")));
        verify(lsdContext, times(2)).capture(message);
    }
}

File: src/test/kotlin/io/lsdconsulting/lsd/distributed/generator/diagram/LsdLoggerShould.kt
package io.lsdconsulting.lsd.distributed.generator.diagram

import com.lsd.core.LsdContext
import com.lsd.core.builders.MessageBuilder.Companion.messageBuilder
import com.lsd.core.domain.Message
import io.lsdconsulting.lsd.distributed.connector.model.InterceptedInteraction
import io.lsdconsulting.lsd.distributed.connector.repository.InterceptedDocumentRepository
import io.lsdconsulting.lsd.distributed.generator.diagram.dto.EventContainer
import io.mockk.every
import io.mockk.mockk
import io.mockk.spyk
import io.mockk.verify
import org.apache.commons.lang3.RandomStringUtils.randomAlphanumeric
import org.jeasy.random.EasyRandom
import org.jeasy.random.EasyRandomParameters
import org.junit.jupiter.api.Test

class LsdLoggerShould {
    private val easyRandom = EasyRandom(EasyRandomParameters().seed(System.currentTimeMillis()))
    private val interceptedDocumentRepository = mockk<InterceptedDocumentRepository>()
    private val interactionGenerator = mockk<InteractionGenerator>()
    private val lsdContext = spyk(LsdContext())
    private val traceId = randomAlphanumeric(8)
    private val secondaryTraceId = randomAlphanumeric(8)
    private val message: Message = messageBuilder().build()

    private val underTest = LsdLogger(interactionGenerator)

    @Test
    fun `capture interaction name`() {
        every { interceptedDocumentRepository.findByTraceIds(traceId) } returns listOf(randomInteraction())
        every { interactionGenerator.generate(any()) } returns EventContainer(events = listOf(message))

        underTest.captureInteractionsFromDatabase(lsdContext, traceId)

        verify { interactionGenerator.generate(mapOf(traceId to null)) }
        verify { lsdContext.capture(message) }
    }

    @Test
    fun `capture interaction names with colour`() {
        every { interceptedDocumentRepository.findByTraceIds(traceId) } returns listOf(randomInteraction())
        every { interceptedDocumentRepository.findByTraceIds(secondaryTraceId) } returns listOf(randomInteraction())
        every { interactionGenerator.generate(any()) } returns EventContainer(events = listOf(message, message))

        underTest.captureInteractionsFromDatabase(
            lsdContext = lsdContext,
            traceIdToColourMap = mapOf(
                traceId to "red",
                secondaryTraceId to "green"
            )
        )

        verify { interactionGenerator.generate(mapOf(traceId to "red", secondaryTraceId to "green")) }
        verify(exactly = 2) { lsdContext.capture(message) }
    }

    private fun randomInteraction(): InterceptedInteraction =
        easyRandom.nextObject(InterceptedInteraction::class.java)
}
